hobbitd/hobbitd_rrd.8
.TH HOBBITD_RRD 8 "Version 4.2.3:  4 Feb 2009" "Xymon"
.SH NAME
hobbitd_rrd \- hobbitd worker module for updating Xymon RRD files
.SH SYNOPSIS
.B "hobbitd_channel --channel=status hobbitd_rrd [options]"
.br
.B "hobbitd_channel --channel=data hobbitd_rrd [options]"

.SH DESCRIPTION
hobbitd_rrd is a worker module for hobbitd, and as such it is normally
run via the
.I hobbitd_channel(8)
program. It receives "status" and "data" messages from hobbitd via
stdin, and updates the RRD databases used to generate trend-graphs.

Clients can send data to Xymon using both status- and data-
messages. So you will normally run two instances of this module, 
once for the "status" channel and once for the "data" channel.

hobbitd_rrd understands data sent by the LARRD 0.43c client-side
scripts (the so-called "bottom-feeder" scripts). So you still want
to install the LARRD bottom-feeders on the clients you monitor.

Note: For certain types of data, the RRD files used by Xymon are
imcompatible with those generated by the Big Brother LARRD add-on. 
See the COMPATIBILITY section below.


.SH OPTIONS
.IP "--debug"
Enable debugging output.

.IP "--rrddir=DIRECTORY"
Defines the directory where the RRD-files are stored. hobbitd_rrd
will use the location pointed to by the BBRRDS environment if this
option is not present.

.IP "--no-cache"
hobbitd_rrd by default caches updates to the RRD files, to reduce
the disk I/O needed for storing the RRD data. Data is collected for
a 30 minute period before being committed to disk in one update.
This option disables caching of the data, so that data is stored
on disk immediately.

.IP "--extra-script=FILENAME"
Defines the script that is run to get the RRD data for tests that are not
built into hobbitd_rrd. You must also specify which tests are handled
by the external script in the \fB--extra-tests\fR option. This option
can only be given once, so the script must handle all of the external
test-data. See the CUSTOM RRD DATA section below. Note that this is
NOT needed if your custom graphs are generated by the NCV (Name Colon
Value) module described below, it is only required for data where you
have a custom script to parse the status message and extract the data
that is put into the graph.

.IP "--extra-tests=TEST[,TEST]"
List of testnames that are handled by the external script. See the
CUSTOM RRD DATA section below. Note that NCV graphs should NOT be
listed here, but in the TEST2RRD environment variable - see below.

.SH ENVIRONMENT
.IP TEST2RRD
Defines the mapping between a status-log columnname and the corresponding
RRD database format. This is normally defined in the 
.I hobbitserver.cfg(5)
file.

.IP BBRRDS
Default directory where RRD files are stored.

.IP NCV_testname
Defines the types of data collected by the "ncv" module in hobbitd_rrd.
See below for more information.

.IP SPLITNCV_testname
The same as NCV_testname, but keeps the data into separate files. That
is, it creates one rrd file per "NAME : value" line found in the
status message. It is useful when the list of NCV lines is varying.

.IP TRACKMAX
Comma-separated list of columnname for which you want to keep the
maximum values along with the default average values. This only works
 for the NCV backend.

.SH COLLECTED DATA
The following RRD-file datasets are generated by hobbitd_rrd:

.IP la
Records the CPU load average. Data is collected from the "cpu"
status report. Requires that a Xymon client is running
on the monitored server.

.IP disk
Records the disk utilization. Data is collected from the "disk"
status report. Requires that a Xymon-compatible client is running
on the monitored server.

.IP memory
Records memory- and swap-utilization. Data is collected from the
"memory" status report. If no "memory" status is reported, it will
use the data from the Win32 client "cpu" status report to generate
this dataset. Requires that a Xymon-compatible client is running
on the monitored server.

.IP netstat
Records TCP and UDP statistics. Data is collected from the "netstat"
status report; however, this data is often sent via the Xymon
"data" protocol, so there need not be a "netstat" column visible
on the Xymon display. To get these data, the LARRD netstat bottom-feeder
script must be running on the monitored server.

.IP vmstat
Records system performance metrics from the "vmstat" command.
Data is collected from the "vmstat" status report; however, this 
data is often sent via the Xymon "data" protocol, so there need 
not be a "vmstat" column visible on the Xymon display. To get 
these data, the LARRD vmstat bottom-feeder script must be running on
the monitored server.

.IP tcp
Response-time metrics from all of the Xymon network tests are
recorded in the "tcp" RRD.

.IP apache
Apache server performance metrics, taken from the "apache" data
report. See the description of the \fBapache\fR keyword in 
.I bb-hosts(5)
for details.

.IP sendmail
Sendmail server performance metrics, taken from the "mailstats"
output. To get these data, the LARRD sendmail bottom-feeder script must
be running on the monitored server.

.IP mailq
Mail queue size. To get these data, the LARRD nmailq bottom-feeder script
must be running on the monitored server.

.IP bea
BEA Weblogic performance data. This is an experimental set of data
collected from BEA Weblogic servers via SNMP, by the "beastats" tool 
included with Xymon.

.IP iishealth
IIS webserver performance data, collected by the "iishealth" script.
This script is a client-side add-on available from the www.deadcat.net 
archive.

.IP temperature
Temperature data, collected with the temperature script from
www.deadcat.net. To get these data, the temperature script must
be running on the monitored server.

.IP ntpstat
Tracks the deviation between the local system time and an NTP
server, using the output from the "ntpq -c rv" command.
A simple script to collect these data is included in the
Xymon contrib/ directory.

.IP citrix
Tracks the number of active sessions on a Citrix server using
the "query session" command. An extension for the BBNT client
that generates data for this graph is in the Xymon contrib/ 
directory.


.SH CUSTOM RRD DATA IN NAME-COLON-VALUE (NCV) FORMAT
Many data-collection scripts report data in the form "NAME : value"
or "NAME = value". So a generic module in hobbitd_rrd allows for
easy tracking of this type of data.

The "ncv" module will automatically detect all occurrences of a "NAME
: value" or "NAME = value" string in a status message, and generate an
RRD file holding all of the name/value data found in the message
(unless you use SPLITNCV, see above). The colon- or equal-sign must be
present - if there is only whitespace, this module will fail.

Only the valid letters (A-Z, a-z) and digits (0-9) are used in the 
dataset names; whitespace and other characters are stripped off 
automatically. Only the first 19 characters of a dataset name are used
(this is an RRD limitation). Underscore '_' is not allowed, even
though RRDtool permits this, and will be stripped from the name.

When using the alternative SPLITNCV_testname, the dataset name is not
limited in length, and non-valid characters are changed to underscores
instead of being stripped off. The dataset inside the resulting rrd
file is always "lambda".

Note that each "NAME : value" must be on a line by itself. If you have
a custom script generating the status- or data-message that is fed
into the NCV handler, make sure it inserts a newline before each
of the data-items you want to track.

To enable the ncv module for a status, add a "COLUMNNAME=ncv" to the 
TEST2RRD setting and the COLUMNNAME to the GRAPHS setting in
.I hobbitserver.cfg(5)
, then restart Xymon. Xymon will now send
all status-messages for the column COLUMNNAME through the hobbitd_rrd
ncv-handler.

The name of the RRD file will be COLUMNNAME.rrd. When using SPLITNCV,
the name of the RRD file will be COLUMNAME,DATASETNAME.rrd.

By default, all of the datasets are generated as the RRD type "DERIVE"
which works for all types of monotonically increasing counters. If you 
have data that are of the type GAUGE, you can override the default via
an environment variable NCV_COLUMNNAME (or SPLITNCV_COLUMNAME). 

E.g. if you are using the bb-mysqlstatus script from www.deadcat.net to 
collect data about your MySQL server, it generates a report in the column 
called "mysql". One data item is the average number of queries/second,
which must be logged in the RRD file as type "GAUGE". To do that, add
the following to hobbitserver.cfg:
.br
    NCV_mysql="Queriespersecondavg:GAUGE" 
.br
If you have multiple datasets that you myst define, add them to the 
environment variable separated by commas, e.g.
.br
    NCV_mysql="Uptime:NONE,Queriespersecondavg:GAUGE" 
.br

The dataset type "NONE" used above causes hobbitd_rrd to ignore this
data, it is not included in the RRD file.

You can use "*" as the dataset name to match all datasets not listed.
E.g.
.br
    NCV_weather="Rain:DERIVE,*:GAUGE"
.br
will cause the "Rainfall" dataset to be of type DERIVE, and all others
of type GAUGE. If you want to track only a few of the variables in
your data, you can use "*:NONE" to drop any dataset not explicitly
listed.

For a more detailed "how to" description, see the on-line HTML
documentation of "How to create graph custom data" available in
the Help menu section on your Xymon server.


.SH CUSTOM RRD DATA VIA SCRIPTS
hobbitd_rrd provides a simple mechanism for adding custom graphs
to the set of data collected on your Xymon server. By adding the
"--extra-script" and "--extra-tests" options, data reported to Xymon
from selected tests are passed to an external script, which can 
define the RRD data-sets to store in an RRD file.

\fBNOTE:\fR For performance reasons, you should not use this mechanism
for large amounts of data. The overhead involved in storing the received
message to disk and launching the script is significantly larger than
the normal hobbitd_rrd overhead. So if you have a large number of
reports for a given test, you should consider implementing it in C
and including it in the hobbitd_rrd tool.

Apart from writing the script, You must also add a section to
.I hobbitgraph.cfg(5)
so that
.I hobbitgraph.cgi(1)
knows how to generate the graph from the data stored in the RRD file. 
To make the graphs actually show up on the status-page and/or the
"trends" page, add the name of the new graph to the TEST2RRD and/or 
GRAPHS setting in
.I hobbitserver.cfg(5).

The script is invoked for each message that arrives, where the test-name
matches one of the testnames given in the "--extra-tests" option. The
script receives three command-line parameters:

.TP
.BI "Hostname"
The name of the host reporting the data.
.TP
.BI "Testname"
The name of the test being reported.
.TP
.BI "Filename"
File containing the data that was reported. This file is generated for
you by hobbitd_rrd, and is also deleted automatically after your script
is finished with it.

.LP
The script must process the data that is reported, and generate the following
output:

.TP
.BI "RRD data-set definitions"
For each dataset that the RRD file holds, a line beginning with "DS:" must be
output.  If multiple data-sets are used, print one line for each dataset.
.br
Data-set definitions are described in the
.I rrdcreate(1)
documentation, but a common definition for e.g. tracking the number of users 
logged on would be "DS:users:GAUGE:600:0:U". "users" is the name of the dataset,
"GAUGE" is the datatype, "600" is the longest time allowed between updates for
the data to be valid, "0" is the minimum value, and "U" is the maximum value
(a "U" means "unknown"). 
.TP
.BI "RRD filename"
The name of the RRD file where the data is stored. Note that Xymon stores all
RRD files in host-specific directories, so unlike LARRD you should not include
the hostname in the name of the RRD file.
.TP
.BI "RRD values"
One line, with all of the data values collected by the script. Data-items are
colon-delimited and must appear in the same sequence as your data-set definitions, 
e.g. if your RRD has two datasets with the values "5" and "0.4" respectively, 
then the script must output "5:0.4" as the RRD values.
.br
In some cases it may be useful to define a dataset even though you will not
always have data for it. In that case, use "U" (unknown) for the value.

If you want to store the data in multiple RRD files, the script can just print out
more sequences of data-set definitions, RRD filenames and RRD values. If the
data-set definitions are identical to the previous definition, you need not
print the data-set definitions again - just print a new RRD filename and value.

.LP
The following sample script for tracking weather data shows how to use this 
mechanism. It assumes the status message include lines like these:
.IP
.nf
green Weather in Copenhagen is FAIR

Temperature: 21 degrees Celsius
Wind: 4 m/s
Humidity: 72 %
Rainfall: 5 mm since 6:00 AM
.fi
.LP
A shell-script to track all of these variables could be written like this:
.IP
.nf
#!/bin/sh

# Input parameters: Hostname, testname (column), and messagefile
HOSTNAME="$1"
TESTNAME="$2"
FNAME="$3"

if [ "$TESTNAME" = "weather" ]
then
	# Analyze the message we got
	TEMP=`grep "^Temperature:" $FNAME | awk '{print $2}'`
	WIND=`grep "^Wind:" $FNAME | awk '{print $2}'`
	HMTY=`grep "^Humidity:" $FNAME | awk '{print $2}'`
	RAIN=`grep "^Rainfall:" $FNAME | awk '{print $2}'`

	# The RRD dataset definitions
	echo "DS:temperature:GAUGE:600:-30:50"
	echo "DS:wind:GAUGE:600:0:U"
	echo "DS:humidity:GAUGE:600:0:100"
	echo "DS:rainfall:DERIVE:600:0:100"

	# The filename
	echo "weather.rrd"

	# The data
	echo "$TEMP:$WIND:$HMTY:$RAIN"
fi

exit 0
.fi


.SH COMPATIBILITY

Some of the RRD files generated by hobbitd_rrd are incompatible with
the files generated by the Big Brother LARRD add-on:

.IP vmstat
The vmstat files with data from Linux based systems are incompatible
due to the addition of a number of new data-items that LARRD 0.43
do not collect, but hobbitd_rrd does. This is due to changes in the
output from the Linux vmstat command, and changes in the way e.g. 
system load metrics are reported.

.IP netstat
All netstat files from LARRD 0.43 are incompatible with hobbitd_rrd.
The netstat data collected by LARRD is quite confusing: For some types
of systems LARRD collects packet-counts, for others it collects byte-
counts. hobbitd_rrd uses a different RRD file-format with separate
counters for packets and bytes and tracks whatever data the system is
reporting.


.SH "SEE ALSO"
hobbitd_channel(8), hobbitd(8), hobbitserver.cfg(5), xymon(7)



Changes
Changes from 4.3.0 beta 1 -> 4.3.0 beta 2 (24 Apr 2009)
=======================================================
* New "--shuffle" option for bbtest-net to run network tests
  in a random order.
* Client startup script now exports important environment variables,
  so they are actually used in systems with a traditional shell.
* hobbitlaunch no longer crashes if there are no tasks
* Client configure script includes librt for clock_gettime()
* New client support for mainframes: z/OS, z/VM, z/VSE
* Enhanced eventlog and top-changing hosts webpages.
* Revert debian package pathnames back to use "hobbit", so
  updates from 4.2.x will actually work.
* Ghostlist options in hobbitcgi.cfg had no effect because
  of typo in setting name.
* "data" messages could crash hobbitd.
* Debug output from hobbitd_channel now logs only the relevant
  data instead of the full message.
* trimhistory now informs the history module to re-open the
  "allevents" file after trimming it.
* devmon template fix
* New "rrdcachectl" utility included.
* Fixed sorting routine (affected holiday list and others)
* Fix generic crash in communications module between Xymon
  programs, where an empty response message would crash caller.

Changes from 4.2.3 -> 4.3.0 beta 1 (09 Feb 2009)
================================================
Core changes:
* New API's for loadhosts and sendmessage, in preparation for
  the full 5.0 changes.
* Always use getcurrenttime() instead of time().
* Support for defining holidays as non-working days in alerts and
  SLA calculations.
* Hosts which appear on multiple pages in the web display can
  use any page they are on in the alerting rules and elsewhere.
* Worker modules (RRD, client-data parsers etc) can operate on
  remote hosts from the hobbitd daemon, for load-sharing.
* Various bugfixes collected over time.

Network test changes:
* Merged new network tests from trunk: SOAP-over-HTTP,
  SSL minimum cipher strength
* Changed network test code to always report a validity period
  for network tests, so it it possible to run network tests less
  often than every 30 minutes (e.g. once an hour).
* Make the content-type setting in HTTP POST tests configurable.
* Make the source-address used for TCP tests configurable.
* Make the acceptable HTTP result codes configurable.
* Use and save HTTP session cookies.

Web changes
* Support generic drop-down lists in templates.
* "NOCOLUMNS" changed to work for all columns.
* New "group-sorted" definition to auto-sort hosts in a group
* Use browser tooltips for host comments
* "Compact" status allows several statuses to appear as a single
  status on the overview webpages.
* Trends page can select the time period to show. Buttons provided
  for the common selections.
* Ghost list report now lists possible candidates for a ghost, 
  based on IP-address or unqualified hostname.

Report changes
* Number of outages as SLA parameter

Miscellaneous
* hobbitlaunch support for running tasks only on certain hosts,
  and for a maximum time.
* Alert script get a unique ID for each alert.



Changes from 4.2.2 -> 4.2.3 (09 Feb 2008)
=========================================
* Time-out code changed to use clock_gettime() with CLOCK_MONOTONIC
* Bugfix for hobbitd/hobbitd_worker communication going out-of-sync
  resulting in "garbled data" being logged and worker modules 
  stopping.
* NCV module now works with negative numbers.
* Several bugfixes in DNS lookup code - could lead to crashes when
  performing DNS tests.
* Switch to C-ARES 1.6.0 - drop support for older versions.
* Run more TCP tests in parallel by not waiting for very slow
  connections to complete before starting new ones.
* Added "hostlist" web utility for spreadsheet-reporting of the
  hosts in Hobbit.


Changes from 4.2.0 -> 4.2.2 (01 Dec 2008)
=========================================

Changelog extracted from Subversion.


------------------------------------------------------------------------
r5935 | storner | 2008-11-26 12:57:11 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   A /branches/4.2.0 (from /trunk:5040)

Create branch for maintenance of the 4.2.x release.


------------------------------------------------------------------------
r5936 | storner | 2008-11-26 13:13:50 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/common/bb-hosts.5

The ability to set a DOWNTIME for an individual test was mentioned in the Changes file, but not documented in the bb-hosts man-page.


------------------------------------------------------------------------
r5937 | storner | 2008-11-26 13:14:48 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitd_client.c

The hobbitd_client program would crash when running in test-mode, i.e. launched with the --test option to test disk-, procs- and ports-settings. This only applies to running it in test-mode, normal operation is not affected.


------------------------------------------------------------------------
r5938 | storner | 2008-11-26 13:15:40 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/lib/headfoot.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/bb-eventlog.c
   M /branches/4.2.0/web/bb-findhost.c
   M /branches/4.2.0/web/bb-rep.c
   M /branches/4.2.0/web/bb-snapshot.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-nkedit.c

Some of the web page would generate an extra line with "Content-type: text/html" at the top of the webpage, although this would most often be hidden by the menubar.


------------------------------------------------------------------------
r5939 | storner | 2008-11-26 13:16:21 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/runclient.sh

Using extra options with the client-side "runclient.sh" script when performing a restart command fails to pass the extra options to the final command. 


------------------------------------------------------------------------

hobbitd/hobbitd_rrd.c
/*----------------------------------------------------------------------------*/
/* Hobbit message daemon.                                                     */
/*                                                                            */
/* This is a hobbitd worker module for the "status" and "data" channels.      */
/* This module maintains the RRD database-files, updating them as new         */
/* data arrives.                                                              */
/*                                                                            */
/* Copyright (C) 2004-2009 Henrik Storner <henrik@hswn.dk>                    */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id$";

#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/time.h>
#include <limits.h>
#include <sys/wait.h>
#include <dirent.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <fcntl.h>
#include <errno.h>

#include "libbbgen.h"
#include "hobbitd_worker.h"

#include "hobbitd_rrd.h"
#include "do_rrd.h"
#include "client_config.h"

#include <signal.h>


#define MAX_META 20	/* The maximum number of meta-data items in a message */

int seq = 0;
static int running = 1;
static time_t reloadtime = 0;

typedef struct rrddeftree_t {
	char *key;
	int count;
	char **defs;
} rrddeftree_t;
static RbtHandle rrddeftree;

static void sig_handler(int signum)
{
	switch (signum) {
	  case SIGHUP:
		  reloadtime = 0;
		  break;
	  case SIGCHLD:
		  break;
	  case SIGINT:
	  case SIGTERM:
		  running = 0;
		  break;
	}
}

static void update_locator_hostdata(char *id)
{
	DIR *fd;
	struct dirent *d;

	fd = opendir(rrddir);
	if (fd == NULL) {
		errprintf("Cannot scan directory %s\n", rrddir);
		return;
	}

	while ((d = readdir(fd)) != NULL) {
		if (*(d->d_name) == '.') continue;
		locator_register_host(d->d_name, ST_RRD, id);
	}

	closedir(fd);
}

static void load_rrddefs(void)
{
	char fn[PATH_MAX];
	FILE *fd;
	strbuffer_t *inbuf = newstrbuffer(0);
	char *key = NULL, *p;
	char **defs = NULL;
	int defcount = 0;
	rrddeftree_t *newrec;

	rrddeftree = rbtNew(name_compare);

	sprintf(fn, "%s/etc/hobbit-rrddefinitions.cfg", xgetenv("BBHOME"));
	fd = stackfopen(fn, "r", NULL);
	if (fd == NULL) goto loaddone;

	while (stackfgets(inbuf, NULL)) {
		sanitize_input(inbuf, 1, 0); if (STRBUFLEN(inbuf) == 0) continue;

		if (*(STRBUF(inbuf)) == '[') {
			if (key && (defcount > 0)) {
				/* Save the current record */
				newrec = (rrddeftree_t *)malloc(sizeof(rrddeftree_t));
				newrec->key = key;
				newrec->defs = defs;
				newrec->count = defcount;
				rbtInsert(rrddeftree, newrec->key, newrec);

				key = NULL; defs = NULL; defcount = 0;
			}

			key = strdup(STRBUF(inbuf)+1);
			p = strchr(key, ']'); if (p) *p = '\0';
		}
		else if (key) {
			if (!defs) {
				defcount = 1;
				defs = (char **)malloc(sizeof(char *));
			}
			else {
				defcount++;
				defs = (char **)realloc(defs, defcount * sizeof(char *));
			}
			p = STRBUF(inbuf); p += strspn(p, " \t");
			defs[defcount-1] = strdup(p);
		}
	}

	if (key && (defcount > 0)) {
		/* Save the last record */
		newrec = (rrddeftree_t *)malloc(sizeof(rrddeftree_t));
		newrec->key = key;
		newrec->defs = defs;
		newrec->count = defcount;
		rbtInsert(rrddeftree, newrec->key, newrec);
	}

	stackfclose(fd);

loaddone:
	freestrbuffer(inbuf);

	/* Check if the default record exists */
	if (rbtFind(rrddeftree, "") == rbtEnd(rrddeftree)) {
		/* Create the default record */
		newrec = (rrddeftree_t *)malloc(sizeof(rrddeftree_t));
		newrec->key = strdup("");
		newrec->defs = (char **)malloc(4 * sizeof(char *));;
		newrec->defs[0] = strdup("RRA:AVERAGE:0.5:1:576");
		newrec->defs[1] = strdup("RRA:AVERAGE:0.5:6:576");
		newrec->defs[2] = strdup("RRA:AVERAGE:0.5:24:576");
		newrec->defs[3] = strdup("RRA:AVERAGE:0.5:288:576");
		newrec->count = 4;
		rbtInsert(rrddeftree, newrec->key, newrec);
	}
}

char **get_rrd_definition(char *key, int *count)
{
	RbtHandle handle;

	handle = rbtFind(rrddeftree, key);
	if (handle == rbtEnd(rrddeftree)) {
		handle = rbtFind(rrddeftree, "");	/* The default record */
	}
	rrddeftree_t *rec = (rrddeftree_t *)gettreeitem(rrddeftree, handle);

	*count = rec->count;
	return rec->defs;
}

int main(int argc, char *argv[])
{
	char *msg;
	int argi;
	struct sigaction sa;
	char *exthandler = NULL;
	char *extids = NULL;
	char *processor = NULL;
	struct sockaddr_un ctlsockaddr;
	int ctlsocket;

	/* Handle program options. */
	for (argi = 1; (argi < argc); argi++) {
		if (strcmp(argv[argi], "--debug") == 0) {
			debug = 1;
		}
		else if (argnmatch(argv[argi], "--rrddir=")) {
			char *p = strchr(argv[argi], '=');
			rrddir = strdup(p+1);
		}
		else if (argnmatch(argv[argi], "--extra-script=")) {
			char *p = strchr(argv[argi], '=');
			exthandler = strdup(p+1);
		}
		else if (argnmatch(argv[argi], "--extra-tests=")) {
			char *p = strchr(argv[argi], '=');
			extids = strdup(p+1);
		}
		else if (argnmatch(argv[argi], "--processor=")) {
			char *p = strchr(argv[argi], '=');
			processor = strdup(p+1);
		}
		else if (strcmp(argv[argi], "--no-cache") == 0) {
			use_rrd_cache = 0;
		}
		else if (net_worker_option(argv[argi])) {
			/* Handled in the subroutine */
		}
	}

	save_errbuf = 0;

	if ((rrddir == NULL) && xgetenv("BBRRDS")) {
		rrddir = strdup(xgetenv("BBRRDS"));
	}

	if (exthandler && extids) setup_exthandler(exthandler, extids);

	/* Do the network stuff if needed */
	net_worker_run(ST_RRD, LOC_STICKY, update_locator_hostdata);

	setup_signalhandler("hobbitd_rrd");
	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = sig_handler;
	sigaction(SIGHUP, &sa, NULL);
	sigaction(SIGCHLD, &sa, NULL);
	sigaction(SIGTERM, &sa, NULL);
	sigaction(SIGINT, &sa, NULL);
	signal(SIGPIPE, SIG_DFL);

	/* Setup the control socket that receives cache-flush commands */
	memset(&ctlsockaddr, 0, sizeof(ctlsockaddr));
	sprintf(ctlsockaddr.sun_path, "%s/rrdctl.%d", xgetenv("BBTMP"), getpid());
	unlink(ctlsockaddr.sun_path);     /* In case it was accidentally left behind */
	ctlsockaddr.sun_family = AF_UNIX;
	ctlsocket = socket(AF_UNIX, SOCK_DGRAM, 0);
	if (ctlsocket == -1) {
		errprintf("Cannot create cache-control socket (%s)\n", strerror(errno));
		return 1;
	}
	fcntl(ctlsocket, F_SETFL, O_NONBLOCK);
	if (bind(ctlsocket, (struct sockaddr *)&ctlsockaddr, sizeof(ctlsockaddr)) == -1) {
		errprintf("Cannot bind to cache-control socket (%s)\n", strerror(errno));
		return 1;
	}
	/* Linux obeys filesystem permissions on the socket file, so make it world-accessible */
	if (chmod(ctlsockaddr.sun_path, S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP|S_IROTH|S_IWOTH) == -1) {
		errprintf("Setting permissions on cache-control socket failed: %s\n", strerror(errno));
	}

	/* Load the RRD definitions */
	load_rrddefs();

	/* If we are passing data to an external processor, create the pipe to it */
	setup_extprocessor(processor);

	running = 1;
	while (running) {
		char *eoln, *restofmsg = NULL;
		char *metadata[MAX_META+1];
		int metacount;
		char *p;
		char *hostname = NULL, *testname = NULL, *sender = NULL, *classname = NULL, *pagepaths = NULL;
		hobbitrrd_t *ldef = NULL;
		time_t tstamp;
		int childstat;
                ssize_t n;
		char ctlbuf[PATH_MAX];
		int gotcachectlmessage;
		time_t now;

		/* See if we have any cache-control messages pending */
		do {
			n = recv(ctlsocket, ctlbuf, sizeof(ctlbuf), 0);
			gotcachectlmessage = (n > 0);
			if (gotcachectlmessage) {
				/* We have a control message */
				char *bol, *eol;

				ctlbuf[n] = '\0';
				bol = ctlbuf;
				do {
					eol = strchr(bol, '\n'); if (eol) *eol = '\0';
					rrdcacheflushhost(bol);
					if (eol) { bol = eol+1; } else bol = NULL;
				} while (bol && *bol);
			}
		} while (gotcachectlmessage);

		/* Get next message */
		msg = get_hobbitd_message(C_LAST, argv[0], &seq, NULL);
		if (msg == NULL) {
			running = 0;
			continue;
		}

		now = gettimer();
		if (reloadtime < now) {
			/* Reload configuration files */
			load_hostnames(xgetenv("BBHOSTS"), NULL, get_fqdn());
			load_client_config(NULL);
			reloadtime = now + 600;
		}

		/* Split the message in the first line (with meta-data), and the rest */
 		eoln = strchr(msg, '\n');
		if (eoln) {
			*eoln = '\0';
			restofmsg = eoln+1;
		}

		/* Parse the meta-data */
		metacount = 0; 
		memset(&metadata, 0, sizeof(metadata));
		p = gettok(msg, "|");
		while (p && (metacount < MAX_META)) {
			metadata[metacount++] = p;
			p = gettok(NULL, "|");
		}
		metadata[metacount] = NULL;

		if ((metacount >= 14) && (strncmp(metadata[0], "@@status", 8) == 0)) {
			/*
			 * @@status|timestamp|sender|origin|hostname|testname|expiretime|color|testflags|\
			 * prevcolor|changetime|ackexpiretime|ackmessage|disableexpiretime|disablemessage|\
			 * clienttstamp|flapping|classname|pagepaths
			 */
			int color = parse_color(metadata[7]);

			switch (color) {
			  case COL_GREEN:
			  case COL_YELLOW:
			  case COL_RED:
			  case COL_BLUE: /* Blue is OK, because it only arrives here when an update is sent */
				tstamp = atoi(metadata[1]);
				sender = metadata[2];
				hostname = metadata[4]; 
				testname = metadata[5];
				classname = (metadata[17] ? metadata[17] : "");
				pagepaths = (metadata[18] ? metadata[18] : "");
				ldef = find_hobbit_rrd(testname, metadata[8]);
				update_rrd(hostname, testname, restofmsg, tstamp, sender, ldef, classname, pagepaths);
				break;

			  default:
				/* Ignore reports with purple, blue or clear - they have no data we want. */
				break;
			}
		}
		else if ((metacount > 5) && (strncmp(metadata[0], "@@data", 6) == 0)) {
			/* @@data|timestamp|sender|origin|hostname|testname|classname|pagepaths */
			tstamp = atoi(metadata[1]);
			sender = metadata[2];
			hostname = metadata[4]; 
			testname = metadata[5];
			classname = (metadata[6] ? metadata[6] : "");
			pagepaths = (metadata[7] ? metadata[7] : "");
			ldef = find_hobbit_rrd(testname, "");
			update_rrd(hostname, testname, restofmsg, tstamp, sender, ldef, classname, pagepaths);
		}
		else if (strncmp(metadata[0], "@@shutdown", 10) == 0) {
			running = 0;
			continue;
		}
		else if (strncmp(metadata[0], "@@idle", 6) == 0) {
			/* Ignored */
			continue;
		}
		else if (strncmp(metadata[0], "@@logrotate", 11) == 0) {
			char *fn = xgetenv("HOBBITCHANNEL_LOGFILENAME");
			if (fn && strlen(fn)) {
				freopen(fn, "a", stdout);
				freopen(fn, "a", stderr);
			}
			continue;
		}
		else if ((metacount > 3) && (strncmp(metadata[0], "@@drophost", 10) == 0)) {
			char hostdir[PATH_MAX];
			hostname = metadata[3];

			MEMDEFINE(hostdir);

			sprintf(hostdir, "%s/%s", rrddir, hostname);
			dropdirectory(hostdir, 1);

			MEMUNDEFINE(hostdir);
		}
		else if ((metacount > 4) && (strncmp(metadata[0], "@@droptest", 10) == 0)) {
			/*
			 * Not implemented. Mappings of testnames -> rrd files is
			 * too complex, so on the rare occasion that a single test
			 * is deleted, they will have to delete the rrd files themselves.
			 */
		}
		else if ((metacount > 4) && (strncmp(metadata[0], "@@renamehost", 12) == 0)) {
			char oldhostdir[PATH_MAX];
			char newhostdir[PATH_MAX];
			char *newhostname;

			MEMDEFINE(oldhostdir);
			MEMDEFINE(newhostdir);

			hostname = metadata[3];
			newhostname = metadata[4];
			sprintf(oldhostdir, "%s/%s", rrddir, hostname);
			sprintf(newhostdir, "%s/%s", rrddir, newhostname);
			rename(oldhostdir, newhostdir);

			if (net_worker_locatorbased()) locator_rename_host(hostname, newhostname, ST_RRD);

			MEMUNDEFINE(newhostdir);
			MEMUNDEFINE(oldhostdir);
		}
		else if ((metacount > 5) && (strncmp(metadata[0], "@@renametest", 12) == 0)) {
			/* Not implemented. See "droptest". */
		}

		/* 
		 * We fork a subprocess when processing drophost requests.
		 * Pickup any finished child processes to avoid zombies
		 */
		while (wait3(&childstat, WNOHANG, NULL) > 0) ;
	}

	/* Flush all cached updates to disk */
	errprintf("Shutting down, flushing cached updates to disk\n");
	rrdcacheflushall();
	errprintf("Cache flush completed\n");

	/* Close the external processor */
	shutdown_extprocessor();

	/* Close the control socket */
	close(ctlsocket);
	unlink(ctlsockaddr.sun_path);

	return 0;
}

